text/2021-04-23-builtin-gc-worker.md

```diff
@@ -0,0 +1,126 @@
+# Builtin GC worker
+
+## Summary
+
+Move TiKV MVCC GC worker from TiDB into PD and make it general for use without TiDB.
+
+## Motivation
+
+GC worker is an important component for TiKV that deletes outdated MVCC data so as to not explode the storage. But currently, the GC worker is implemented in TiDB, which makes TiKV not usable without TiDB.
+
+## Background
+
+According to the [documentation](https://docs.pingcap.com/tidb/stable/garbage-collection-overview) for the **current** GC worker, the GC worker processes the following steps in a GC cycle:
+
+1. Calculate the minimal timestamp among all living SQL sessions, earlier than which are safe to GC, called `Service Safepoint`.
+2. [Upload](https://github.com/pingcap/kvproto/blob/8ecb5e46d7f5f7952a1a8d262b54f61dc8de1ef3/proto/pdpb.proto#L73) the `Service Safepoint` to PD, and get the minimal `Service Safepoint` among all Tools (e.g. CDC) from the response.
+3. Resolve all locks whose timestamp is earlier than the `Minimal Service Safepoint`.
+4. [Push](https://github.com/pingcap/kvproto/blob/8ecb5e46d7f5f7952a1a8d262b54f61dc8de1ef3/proto/pdpb.proto#L71) the `Minimal Service Safepoint` as `GCSafepoint` to PD.
+
+After the `GCSafepoint` being uploaded, TiKV will pull it from PD by interval, then TiKV will automatically delete all records earlier than `GCSafepoint`.
+
+Yet another tricky part is that TiDB has a shortcut to GC a range of keys after `DROP TABLE` or `DROP INDEX`: parallel with step 3 and step 4, TiDB will directly invoke `UnsafeDeleteRange` to TiKV which will delete the range of key-values regardless of locks or timestamp. To make it safe, TiDB will guarantee that the deleted range will never be used again.
+
+## Detailed design
+
+Three new concepts will be introduced by this proposal:
+
+1. `GC Barrier`
+
+    PD guarantees that the GC process will not affect the data with a timestamp later than any `GC Barrier`. TiKV users (e.g. TiDB / CDC / Client) can put GC Barriers into PD so as to make sure the data they are using will not be GC.
+
+2. `Transaction Safepoint`
+
+    TiKV and TiFlash guarantee that no read or write will happen on the data earlier than `Transaction Safepoint`. PD calculates `Transaction Safepoint Proposal` by `min(GC Barriers) - 1` and proposes it to TiKV and TiFlash via the `StoreHeartbeatResponse`. Therefore, TiKV and TiFlash will reject all requests with timestamp earlier than the `Transaction Safepoint`. In the next following `StoreHeartbeatRequest` TiKV and TiFlash will report their latest `Transaction Safepoint Status` to PD.
```

Comment:
Why do we need this concept? IMO, both TiFlash and TiKV shouldn't care about validity of timestamp.

Reply:
This is for solving a very tricky corner case:

Reply:
> `A region on StoreB having MVCC data only later than ts=100, anyhow transferred to StoreA via raft snapshot.`

This decision is made by PD. It has the choice that not to do this in the situation where Involved tikvs have different safepoints.

Reply:
After all, "it's PD's fault". 🤣

Reply:
But I do agree stores should reject requests with stale ts. There should be something wrong to encounter these requests sent by some client. I think it's better to report this danger to the client.

Reply:
> But I do agree stores should reject requests with stale ts. There should be something wrong to encounter these requests sent by some client. I think it's better to report this danger to the client.

Rejection of request with stale ts can prevent extra resource consumption either.

Reply:
The mechanism described here adds extra complexity while not solve the problem completely. For example, an ongoing read right before transaction safepoint can be corrupted by unsafe delete range later. The behavior of reading data before safepoint is undefined. Client should make sure it's operating on the safe data, for example, making sure read timestamp is larger than the GC safepoint, by not advancing the GC Barrier.